x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/TransformClusterStateListener.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.transform;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.admin.indices.alias.IndicesAliasesRequest;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.gateway.GatewayService;
+import org.elasticsearch.xpack.core.transform.transforms.persistence.TransformInternalIndexConstants;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.elasticsearch.xpack.core.ClientHelper.TRANSFORM_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+
+class TransformClusterStateListener implements ClusterStateListener {
+
+    private static final Logger logger = LogManager.getLogger(TransformClusterStateListener.class);
+
+    private final Client client;
+    private final AtomicBoolean isIndexCreationInProgress = new AtomicBoolean(false);
+
+    TransformClusterStateListener(ClusterService clusterService, Client client) {
+        this.client = client;
+        clusterService.addListener(this);
+        logger.debug("Created TransformClusterStateListener");
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+            // Wait until the gateway has recovered from disk.
+            return;
+        }
+
+        // The atomic flag prevents multiple simultaneous attempts to run alias creation
+        // if there is a flurry of cluster state updates in quick succession
+        if (event.localNodeMaster() && isIndexCreationInProgress.compareAndSet(false, true)) {
+            createAuditAliasForDataFrameBWC(event.state(), client, ActionListener.wrap(
+                r -> {
+                    isIndexCreationInProgress.set(false);
```

Comment:
Should this statement be the last in this function?

Reply:
I think it's safer that it's first.  If this line of code doesn't get run because another line in the response handler throws an exception then the only way another attempt at alias creation will be made is if the node is restarted.  (Obviously this is a bigger problem for the failure case than the success case.)  In the failure case it doesn't matter than a retry kicks off while a log message about the first failure is still being created.  The main thing is that we don't end up with 100 parallel attempts to create the alias as this blocks up the cluster state update thread.